### PR TITLE
split cluster policy from local policy

### PR DESCRIPTION
**Pull request by deads2k**
External access always deals with ClusterPolicy and Policy, while internal access deals in Policy.  This can be a little confusing to read, but at this stage it seems more important to get the data right than the impl beautiful.

@derekwaynecarr can you review?

### CODE REVIEW

**Comment by deads2k** [comment]
[test]

**Comment by deads2k** [comment]
Updated the project cache to handle cluster policies.  I think it could use some refactoring (looks gorpy), but you'll probably get to that when dealing with the policy cache race.

**Comment by derekwaynecarr** [comment]
this PR is huge, will look to review later this evening.

**Comment by deads2k** [comment]
> this PR is huge, will look to review later this evening.

This is less than half the size of the next best solution.  If you can make a smaller one.... :)

File: hack/test-cmd.sh [hunk 1]
```diff
@@ -460,8 +464,8 @@ osadm policy add-role-to-user admin adduser -n ui-test-project
 # Make sure project can be listed by osc (after auth cache syncs)
 sleep 2 && [ "$(osc get projects | grep 'ui-test-project')" ]
 # Make sure users got added
-[ "$(osc describe policybinding master:default -n ui-test-project | grep createuser)" ]
```
Comment (liggitt) [inline hack/test-cmd.sh]: is the leading colon needed? wouldn't something without a namespace component be a cluster binding?

Comment (liggitt) [inline hack/test-cmd.sh]: unless we don't prevent `:` in the policy name...

File: pkg/authorization/authorizer/authorizer_test.go [hunk 2]
```diff
@@ -430,45 +432,40 @@ func matchError(expected string, actual error, field string, t *testing.T) {
...
+	return []authorizationapi.ClusterPolicy{*GetBootstrapPolicy()}
 }
-func newDefaultGlobalBinding() []authorizationapi.PolicyBinding {
-	policyBinding := authorizationapi.PolicyBinding{
+func newDefaultGlobalBinding() []authorizationapi.ClusterPolicyBinding {
+	policyBinding := authorizationapi.ClusterPolicyBinding{
 		ObjectMeta: kapi.ObjectMeta{
-			Name:      bootstrappolicy.DefaultMasterAuthorizationNamespace,
-			Namespace: bootstrappolicy.DefaultMasterAuthorizationNamespace,
+			Name: authorizationapi.GetPolicyBindingName(""),
```
Comment (liggitt) [inline pkg/authorization/authorizer/authorizer_test.go]: use `MasterPolicyBindingName`?

File: pkg/authorization/authorizer/authorizer_test.go [hunk 3]
```diff
@@ -497,7 +494,7 @@ func newAdzeBindings() []authorizationapi.PolicyBinding {
 	return []authorizationapi.PolicyBinding{
 		{
 			ObjectMeta: kapi.ObjectMeta{
-				Name:      bootstrappolicy.DefaultMasterAuthorizationNamespace,
+				Name:      authorizationapi.GetPolicyBindingName(""),
```
Comment (liggitt) [inline pkg/authorization/authorizer/authorizer_test.go]: use `MasterPolicyBindingName`?

File: pkg/authorization/authorizer/bootstrap_policy_test.go [hunk 4]
```diff
@@ -451,7 +451,7 @@ func newMalletBindings() []authorizationapi.PolicyBinding {
 	return []authorizationapi.PolicyBinding{
 		{
 			ObjectMeta: kapi.ObjectMeta{
-				Name:      bootstrappolicy.DefaultMasterAuthorizationNamespace,
+				Name:      authorizationapi.GetPolicyBindingName(""),
```
Comment (liggitt) [inline pkg/authorization/authorizer/bootstrap_policy_test.go]: `MasterPolicyBindingName`?

File: pkg/authorization/registry/rolebinding/policybased/virtual_storage_test.go [hunk 5]
```diff
@@ -55,11 +36,43 @@ func testNewBasePolicies() []authorizationapi.Policy {
 	}
 }
 
+func testNewClusterBindings() []authorizationapi.ClusterPolicyBinding {
+	return []authorizationapi.ClusterPolicyBinding{
+		{
+			ObjectMeta: kapi.ObjectMeta{Name: authorizationapi.GetPolicyBindingName("")},
```
Comment (liggitt) [inline pkg/authorization/registry/rolebinding/policybased/virtual_storage_test.go]: `MasterPolicyBindingName`?

File: pkg/authorization/registry/subjectaccessreview/rest_test.go [hunk 6]
```diff
@@ -67,7 +66,7 @@ func TestNoErrors(t *testing.T) {
 			reason:  "because good things",
 		},
 		reviewRequest: &authorizationapi.SubjectAccessReview{
-			Groups:   util.NewStringSet(bootstrappolicy.DefaultMasterAuthorizationNamespace),
+			Groups:   util.NewStringSet("not-master"),
```
Comment (liggitt) [inline pkg/authorization/registry/subjectaccessreview/rest_test.go]: that was a weird place to use `DefaultMasterAuthorizationNamespace` before

File: pkg/authorization/registry/clusterpolicy/registry.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+	GetClusterPolicy(ctx kapi.Context, id string) (*authorizationapi.ClusterPolicy, error)
+	// CreateClusterPolicy creates a new policy.
+	CreateClusterPolicy(ctx kapi.Context, policy *authorizationapi.ClusterPolicy) error
+	// UpdateClusterPolicy updates a policy.
+	UpdateClusterPolicy(ctx kapi.Context, policy *authorizationapi.ClusterPolicy) error
+	// DeleteClusterPolicy deletes a policy.
+	DeleteClusterPolicy(ctx kapi.Context, id string) error
+}
+
+type WatchingRegistry interface {
```
Comment (liggitt) [inline pkg/authorization/registry/clusterpolicy/registry.go]: why the separate type?

File: pkg/authorization/registry/clusterrolebinding/proxy/proxy.go [hunk 8]
```diff
@@ -7,17 +7,31 @@ import (
...
 type ClusterRoleBindingStorage struct {
-	masterNamespace    string
-	typeConverter      authorizationapi.TypeConverter
-	roleBindingStorage rolebindingregistry.Storage
+	roleBindingStorage rolebindingstorage.VirtualStorage
 }
 
-func NewClusterRoleBindingStorage(masterNamespace string, roleBindingStorage rolebindingregistry.Storage) *ClusterRoleBindingStorage {
-	return &ClusterRoleBindingStorage{masterNamespace, authorizationapi.TypeConverter{masterNamespace}, roleBindingStorage}
+func NewRoleBindingStorage(clusterPolicyRegistry clusterpolicyregistry.Registry, clusterBindingRegistry clusterpolicybindingregistry.Registry) *ClusterRoleBindingStorage {
```
Comment (liggitt) [inline pkg/authorization/registry/clusterrolebinding/proxy/proxy.go]: func name `NewClusterRoleBindingStorage`?

File: pkg/authorization/registry/clusterpolicy/etcd/etcd.go [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/generic"
+	etcdgeneric "github.com/GoogleCloudPlatform/kubernetes/pkg/registry/generic/etcd"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/tools"
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+	"github.com/openshift/origin/pkg/authorization/registry/clusterpolicy"
+)
+
+const ClusterPolicyPath = "/registry/authorization/clusterpolicy"
```
Comment (liggitt) [inline pkg/authorization/registry/clusterpolicy/etcd/etcd.go]: I thought we had dropped the registry prefix

File: hack/test-cmd.sh (hunk 1 above)
Comment (deads2k) [inline hack/test-cmd.sh]: > is the leading colon needed? wouldn't something without a namespace component be a cluster binding?

Policy name is always `default`.  The format for `PolicyBinding.Name` is `policy.Namespace:policy.Name`.  Namespaces can't contain ':', so I chose it as a delimiter.  When you have a `PolicyBinding` pointing to a `ClusterPolicy`, you get an empty namespace for the `policy.Namespace`.

File: pkg/authorization/registry/subjectaccessreview/rest_test.go (hunk 6 above)
Comment (deads2k) [inline pkg/authorization/registry/subjectaccessreview/rest_test.go]: > that was a weird place to use DefaultMasterAuthorizationNamespace before

Who reviewed that! :)

File: pkg/authorization/registry/clusterpolicy/registry.go (hunk 7 above)
Comment (deads2k) [inline pkg/authorization/registry/clusterpolicy/registry.go]: > why the separate type?

To keep the internal code coherent, `ClusterPolicy` has an adapting registry that can make the it appear as a `Policy.Registry` **except** for watches, which are really hard to adapt.  I turned out that most of the code that wanted to deal with both scope of `Policy` and didn't care which was which (all of the authorization code) didn't use the watch.  Splitting into a separate interface allowed for the proxying registry.

Given a proper "all" value for a namespace, there is no ambiguity about whether a request is for something without a namespace, something with a namespace, and something for all namespaces.  That made the behavior of the proxying registry consistent and would make a unified registry possible, (if we ever kill the three meanings of namespace="").

File: pkg/authorization/registry/clusterpolicy/etcd/etcd.go (hunk 9 above)
Comment (deads2k) [inline pkg/authorization/registry/clusterpolicy/etcd/etcd.go]: > I thought we had dropped the registry prefix. Maybe I'm getting ahead of myself and am thinking of upstream

Ahead of yourself.

**Comment by deads2k** [comment]
comments addressed.

**Comment by deads2k** [comment]
rebased

**Comment by deads2k** [comment]
@derekwaynecarr bump

**Comment by deads2k** [comment]
comments addressed so far.

**Comment by derekwaynecarr** [comment]
phew ;-)  few more nits and questions.

**Comment by deads2k** [comment]
comments addressed.

**Comment by derekwaynecarr** [comment]
LGTM

File: pkg/project/auth/cache.go [hunk 10]
```diff
@@ -242,6 +281,36 @@ func (ac *AuthorizationCache) synchronizePolicyBindings(userSubjectRecordStore c
...
+// synchronizePolicies synchronizes access over each policy
+func (ac *AuthorizationCache) synchronizeClusterPolicies(userSubjectRecordStore cache.Store, groupSubjectRecordStore cache.Store, reviewRecordStore cache.Store) {
+	items := ac.clusterPolicyIndexer.List()
+	for i := range items {
+		policy := items[i].(*authorizationapi.ClusterPolicy)
+		reviewRequest := &reviewRequest{
+			namespace:                  policy.Namespace,
+			policyUIDToResourceVersion: map[types.UID]string{policy.UID: policy.ResourceVersion},
+		}
+		if err := ac.syncHandler(reviewRequest, userSubjectRecordStore, groupSubjectRecordStore, reviewRecordStore); err != nil {
```
Comment (derekwaynecarr) [inline pkg/project/auth/cache.go]: I now see no point in doing this operation, or anything in this function.  The whole point to synchronize against these is to get an updated reviewRecord for a specific namespace.  Cluster policies are not in a namespace, so calling syncRequest for a no named namespace is meaningless.  Get rid of this.

File: pkg/project/auth/cache.go [hunk 11]
```diff
@@ -242,6 +281,36 @@ func (ac *AuthorizationCache) synchronizePolicyBindings(userSubjectRecordStore c
...
+			util.HandleError(fmt.Errorf("error synchronizing: %v", err))
+		}
+	}
+}
+
+// synchronizePolicyBindings synchronizes access over each policy binding
+func (ac *AuthorizationCache) synchronizeClusterPolicyBindings(userSubjectRecordStore cache.Store, groupSubjectRecordStore cache.Store, reviewRecordStore cache.Store) {
```
Comment (derekwaynecarr) [inline pkg/project/auth/cache.go]: Get rid of this as well.  It has no purpose.  We get no benefit finding out if the user has access to the empty string namespace.

**Comment by deads2k** [comment]
@derekwaynecarr  Pulled the two methods and the block.  Squashed before I realized you'd changed your mind.  Mind checking those final three in place?

**Comment by deads2k** [comment]
comments addressed.

**Comment by derekwaynecarr** [comment]
LGTM, ideally, I would like a comment stating that invalidateCache makes an assumption that there is a single cluster* resource, and maybe even I would like a log entry if we find more than one that behavior will be indeterminate... because right now if you had more than one, it will always invalidate and rebuild entire cache...

**Comment by deads2k** [comment]
comment clarified.  [merge]

**Pull request merged**